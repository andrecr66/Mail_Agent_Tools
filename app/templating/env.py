from __future__ import annotations
from pathlib import Path
from typing import Any, Dict
from jinja2 import Environment, FileSystemLoader, select_autoescape


def _paragraphize(text: str) -> str:
    """Split on blank lines into <p> blocks; convert single newlines to <br/>."""
    parts = [p.strip() for p in text.split("\n\n") if p.strip()]
    return "".join(f"<p>{p.replace('\n', '<br/>')}</p>" for p in parts) or ""


def jinja_env(templates_root: str | Path = "templates/jinja") -> Environment:
    env = Environment(
        loader=FileSystemLoader(str(templates_root)),
        autoescape=select_autoescape(enabled_extensions=("html", "xml")),
        trim_blocks=True,
        lstrip_blocks=True,
    )
    env.filters["paragraphize"] = _paragraphize
    return env


def render_template(template_path: str, context: Dict[str, Any]) -> str:
    env = jinja_env()
    tpl = env.get_template(template_path)
    return tpl.render(**context)
"""Jinja2 environment helpers.

Provides a small set of helpers for rendering HTML emails using Jinja2,
including a `paragraphize` filter that turns newline-separated text into
HTML paragraphs and <br/> line breaks.
"""
