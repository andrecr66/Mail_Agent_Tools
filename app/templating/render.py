# mypy: disable-error-code=import-untyped
from __future__ import annotations
from typing import Any, Dict, Tuple, cast

from bs4 import BeautifulSoup
from premailer import transform

from app.tools.brand_loader import load_brand
from app.templating.env import render_template, jinja_env


def derive_preheader(plain_text: str, limit: int = 90) -> str:
    s = " ".join(plain_text.split())
    return s[:limit]


def to_plain_text(html: str) -> str:
    soup = BeautifulSoup(html, "lxml")
    txt: str = soup.get_text(separator=" ", strip=True)
    return " ".join(txt.split())


def inline_css(html: str) -> str:
    # premailer returns Any to mypy; cast to str for our contract
    return cast(str, transform(html, disable_validation=True))


def render_generic_email(
    *,
    subject: str,
    body_text: str,
    brand_id: str = "default",
    purpose: str = "generic",
    variables: Dict[str, Any] | None = None,
) -> Tuple[str, str]:
    """
    Render the generic_v1 Jinja template with a brand and content variables.
    Returns (html_inlined, plaintext).
    """
    brand = load_brand(brand_id)
    vars = variables or {}
    preheader = vars.get("preheader") or derive_preheader(body_text)
    cta_text = vars.get("cta_text") or "Learn more"
    cta_url = vars.get("cta_url") or brand.links.get("website") or "#"

    # Re-render dynamic footer/signature strings as Jinja templates
    env = jinja_env()
    footer_html = brand.footer_html
    signature_html = brand.signature_html
    if footer_html:
        footer_html = env.from_string(footer_html).render(
            brand=brand, subject=subject, body_text=body_text, purpose=purpose, **vars
        )
    if signature_html:
        signature_html = env.from_string(signature_html).render(
            brand=brand, subject=subject, body_text=body_text, purpose=purpose, **vars
        )

    context = {
        "brand": brand,
        "subject": subject,
        "preheader": preheader,
        "body_text": body_text,
        "cta_text": cta_text,
        "cta_url": cta_url,
        "purpose": purpose,
        "footer_html": footer_html,
        "signature_html": signature_html,
    }
    raw_html = render_template("families/generic/generic_v1.html.j2", context)
    html = inline_css(raw_html)
    text = to_plain_text(html)
    return html, text
