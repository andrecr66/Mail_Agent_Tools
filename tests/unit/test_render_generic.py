from __future__ import annotations
from typing import Any
from app.templating.render import render_generic_email


def test_generic_v1_html_snapshot(snapshot: Any) -> None:
    html, text = render_generic_email(
        subject="Welcome to CodeRoad",
        body_text="Hello!\n\nThis is a short paragraph.\nAnd a second line.",
        brand_id="default",
        purpose="welcome",
        variables={"cta_text": "Visit CodeRoad", "cta_url": "https://coderoad.com/"},
    )
    assert html == snapshot
    assert len(text) > 10
