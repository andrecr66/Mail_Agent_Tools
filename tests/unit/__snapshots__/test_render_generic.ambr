# serializer version: 1
# name: test_generic_v1_html_snapshot
  '''
  <html>
  <head>
    <meta charset="utf-8">
    <title>Welcome to CodeRoad</title>
    <meta name="x-preheader" content="Hello! This is a short paragraph. And a second line.">
    </head>
  <body style="background:#FFF; color:#111827; font-family:Arial, Helvetica, sans-serif">
  <table role="presentation" width="100%" cellpadding="0" cellspacing="0" style="margin:0 auto;max-width:600px;">
    <tr>
      <td style="padding:24px 16px;text-align:left;">
          <div style="font-weight:700;font-size:18px;color:#6B21A8;">CodeRoad</div>
      </td>
    </tr>
  </table>
    <table role="presentation" width="100%" cellpadding="0" cellspacing="0" style="margin:0 auto;max-width:600px;">
      <tr>
        <td>
          <div class="card" style="background:#fff; margin:0 auto; max-width:600px; border-radius:8px; padding:24px 16px">
            <p>Hello!</p><p>This is a short paragraph.<br>And a second line.</p>
  <table role="presentation" cellpadding="0" cellspacing="0" style="margin:16px 0;">
    <tr>
      <td style="background:#6B21A8;border-radius:6px;">
        <a href="https://coderoad.com/" style="color:#fff; display:inline-block; padding:12px 16px; text-decoration:none; font-weight:600">
          Visit CodeRoad
        </a>
      </td>
    </tr>
  </table>        </div>
        </td>
      </tr>
    </table>
  
  <table role="presentation" width="100%" cellpadding="0" cellspacing="0" style="margin:0 auto;max-width:600px;">
    <tr>
      <td style="padding:24px 16px;color:#6b7280;font-size:12px;">
        <p>Â© CodeRoad</p>
      </td>
    </tr>
  </table></body>
  </html>
  '''
# ---
